### ISSUE TITLE
Use of omp_set_dynamic() creates nonsensical OpenMP behaviour, using only 1 thread

### ISSUE THREAD

**Issue by nh2** [opencv/opencv#25717]
### System Information

All Linux systems, probably more, see `libomp` code.

### Detailed description

@alalek I think the behaviour introduced in

* https://github.com/opencv/opencv/pull/12796

is bad because of how [`OMP_DYNAMIC`] is implemented. It makes that OpenMP does not use all available cores.

I noticed that when starting my C++ app, only 1 thread is used, even though the system is idle.

* On other machines with the same spec, running the same binary, all cores were used.
* Running the app a few minutes later made a couple more threads used.
* The more I waited, the more performant the app got.

Setting e.g. `OMP_NUM_THREADS=100` did not help.

I took a wild guess and observed the observed parallelism against the 15-minute load average of the machine:

```
15 min load average
    2.55 -> 10 threads
    4.64 -> 8 threads
    4.93 -> 7 threads
```

My program's code path in question doesn't even use, but only link/load OpenCV.

I eventually found the implementation: [`libgomp/config/linux/proc.c`](https://github.com/gcc-mirror/gcc/blob/10cb3336ba1ac89b258f627222e668b023a6d3d4/libgomp/config/linux/proc.c#L180-L188)

```c
/* When OMP_DYNAMIC is set, at thread launch determine the number of
   threads we should spawn for this team.  */
/* ??? I have no idea what best practice for this is.  Surely some
   function of the number of processors that are *still* online and
   the load average.  Here I use the number of processors online
   minus the 15 minute load average.  */

unsigned
gomp_dynamic_max_threads (void) {
///
    return n_onln - loadavg;
```

## This is really bad

* Because of this logic, your app will use only 1 thread, even though the system is completely idle _now_, just because it was busy 10 minutes ago.
* The dynamic limit is determined _at process start_ (loading time), and fixed forever.
  * So started programs stay slow _forever_ if they were started at a time 5 minutes after the system was busy.
* It means a server can never achieve full utilisation when working down a queue of jobs.
  * Say you have 8 cores, and a queue of N jobs to process, each of which takes 15 minutes full-CPU.
  * The first jobs starts at 0 15-min-utilisation, thus using all cores.
  * The next job starts, using only 1 core.
  * The next job starts, using only 7 cores.
  * The next job starts, using only 1 cores.
  * The next job starts, using only 7 cores.
  * ...
  * In the long run, the **server uses only half of its cores on average**.
* It makes performance behaviour completely irreproducible.
* This should not be on by default.

OpenCV isn't really to blame here, because the [docs of `OMP_DYNAMIC`](https://gcc.gnu.org/onlinedocs/libgomp/OMP_005fDYNAMIC.html) point none of this out. You have to read the implementation to learn it.

## Solution

I petition OpenCV to disable this logic by default, setting `OPENCV_FOR_OPENMP_DYNAMIC_DISABLE=1` by default.

### Steps to reproduce

```sh
stress-ng --cpu 0
# wait 15 minutes, then Ctrl+C

# Now run your app that links OpenCV. 
# It will use only 1 thread for the next 10 minutes.
```

### Issue submission checklist

- [X] I report the issue, it's not a question
- [X] I checked the problem with documentation, FAQ, open issues, forum.opencv.org, Stack Overflow, etc and have not found any solution
- [X] I updated to the latest OpenCV version and the issue is still there
- [X] There is reproducer code and related data files (videos, images, onnx, etc)

**Comment by opencv-alalek** [comment]
Agreed that `libgomp` implementation looks strange. And described behavior is not that users really want.

**Comment by asmorkalov** [comment]
Looks like the original `omp_set_dynamic()` call is a mistake and it should be replaced with `omp_set_schedule(omp_sched_dynamic)`: https://www.openmp.org/spec-html/5.0/openmpsu121.html. OpenCV relies on dynamic dispatching like TBB does. It's the closest behaviour we need.

**Comment by nh2** [comment]
> it should be replaced with `omp_set_schedule(omp_sched_dynamic)`

@asmorkalov OpenCV is a library, I'm not sure it should set any of the global OpenMP options at all.

If a loop in OpenCV benefits from dynamic scheduling, shouldn't that option be set on that loop?

**Comment by nh2** [comment]
Thanks for the fix!

**Comment by mstebelev** [comment]
Some related issue in pytorch https://github.com/pytorch/pytorch/issues/168965

**Comment by nh2** [comment]
Small update:

I also posted an answer here: https://stackoverflow.com/questions/78584145/why-does-my-openmp-app-sometimes-use-only-1-thread-sometimes-3-sometimes-all-c/78584146#78584146

And a bug for `libgomp`:

* https://gcc.gnu.org/bugzilla/show_bug.cgi?id=115367
  * `The implementation of OMP_DYNAMIC is not dynamic `

**Comment by nh2** [comment]
I also just re-discovered this problem via another path: Using Python [numba](https://github.com/numba/numba) I had a problem that it didn't use all cores on the system, leading to low multi-threaded %CPU utilisation and imbalanced `prange` ranges with `@numba.njit(parallel=True)` (numba also uses its `omp` OpenMP backend by default). I discovered that this happens in relation to `import cv2`, in a semi-nondeterministic way. Then I remembered this issue.

I filed that here:

* https://github.com/numba/numba/issues/10454
